### ISSUE TITLE
benchmark cache persist

### ISSUE THREAD

**Issue by jiangzzsss** [pytorch/pytorch#88410]
### 🚀 The feature, motivation and pitch

We've been working on a detection model's training and when we set torch.backends.cudnn.benchmark = True and do some padding , the training speed can improve a lot . 
However if we skip the padding, since the params of convolution vary frequently(e.g. input shapes), it  becomes even slower. 
We are wondering if we can  persist the benchmark cache, which is a map and stores mapping from ConvolutionParams to convolution algorithm. The first person can store some cache and later others will no longer have to suffer from the algorithms choosing overhead.

Will this feature be possible in the future or is already ongoing ? 
**Or there are limits that this whole persisting thing is not a good idea. **

### Alternatives

For example, maybe we can define load() and save() functions inside the global variables:
```
BenchmarkCache<cudnnConvolutionFwdAlgoPerf_t> fwd_algos;
BenchmarkCache<cudnnConvolutionBwdDataAlgoPerf_t> bwd_data_algos;
BenchmarkCache<cudnnConvolutionBwdFilterAlgoPerf_t> bwd_filter_algos;
``` 
then we have to determine the time of load and save are executed, a draft would be like this:
```
template <typename T>
struct BenchmarkCache {
  std::mutex mutex;
  std::unordered_map<ConvolutionParams, T, ParamsHash<ConvolutionParams>, ParamsEqual<ConvolutionParams>> map;
  std::string CACHE_STORED_PATH;
  BenchmarkCache(std::string cache_stored_path): CACHE_STORED_PATH(cache_stored_path){}

  bool find(const ConvolutionParams& params, T* results) {
    std::lock_guard<std::mutex> guard(mutex);
    auto it = map.find(params);
    if (it == map.end()) {
      return false;
    }
    *results = it->second;
    return true;
  }

  void insert(const ConvolutionParams& params, const T& results) {
    std::lock_guard<std::mutex> guard(mutex);
    map[params] = results;
  }

  int load(const &std::string CACHE_STORE_PATH){
    std::lock_guard<std::mutex> guard(mutex);
    //load the cache;
  }

  int save(const &std::string CACHE_STORE_PATH){
    std::lock_guard<std::mutex> guard(mutex);
    //save the cache;
  }
};
```

**Comment by Jack47** [comment]
any updates on this? we would like to contribute this feature if possible. The may issues are:

1. when to execute load once and only once? is it a good idea to load when BenchmarkCache constructs?
2. when to execute save? when Benchmark destructs? 

Because loading/saving file consumes time and exceptions may happen, so is there any pytorch lifecycle hooks to do those init things?

**Comment by ngimel** [comment]
We will accept a PR adding this. As for the question, lazily loading it the first time benchmarking is requested would make sense. As for saving, adding to the cache every time a benchmark is computed is also not a big deal, given that benchmarking is very slow and causes cuda synchronization anyway. You shouldn't cause exception though, if writing doesn't succeed you should just abandon it. 
Another thing to keep in mind, cache key should also contain cuda version, cudnn version and hardware compute capability and memory.
Also, provide a documented way to delete the cache.

**Comment by Jack47** [comment]
memory means gpu memory capacity? it's used to distinguish same compute capability with different memory capacity? such as A100 40G vs 80G? we will test to see whether the best conv algo is same in this situation.

Great thanks for your cache key advice!

**Comment by jiangzzsss** [comment]
Another question is that we notice the cache key is stored in the benchmark_cache.
For efficiency and memory or disk budget, should we imitate the ParamsHash(cudnn 7) to write a function like:
```
template <typename Params>
size_t func(const Params& params) const {
    auto ptr = reinterpret_cast<const uint8_t*>(&params);
    uint32_t value = 0x811C9DC5;
    for (int i = 0; i < (int)sizeof(Params); ++i) {
      value ^= ptr[i];
      value *= 0x01000193;
    }
    return (size_t)value;
  }
```
then use the calculated value instead, so the map would be：
```
std::unordered_map<size_t, T> map;
```
**Or this is actually not necessary ？**

**Comment by ngimel** [comment]
Even for data parallel models, systems sometimes have different GPUs, so discarding ids is not safe. If we could establish that GPUs are the same cc and same memory, probably we could discard ids in that case? cc @eqy.

**Comment by eqy** [comment]
If we don't need to share caches across nodes, there might be other fields e.g., the PCI-E bus ID or GPU name that would be relevant? 
https://developer.download.nvidia.com/compute/DevZone/docs/html/C/doc/html/group__CUDART__DEVICE_g5aa4f47938af8276f08074d09b7d520c.html
